squeal-postgresql/src/Squeal/PostgreSQL/Expression/Time.hs

```diff
@@ -133,41 +135,37 @@ makeTimestamptz = unsafeFunctionN "make_timestamptz"
 
->>> printSQL $ dateTrunc ("day" *: makeTimestamp (2010 :* 5 :* 6 :* 14 :* 45 *: 11.4))
-date_trunc((E'day' :: text), make_timestamp((2010 :: int4), (5 :: int4), (6 :: int4), (14 :: int4), (45 :: int4), (11.4 :: float8)))
+>>> printSQL $ dateTrunc Quarter (makeTimestamp (2010 :* 5 :* 6 :* 14 :* 45 *: 11.4))
+date_trunc('quarter', make_timestamp((2010 :: int4), (5 :: int4), (6 :: int4), (14 :: int4), (45 :: int4), (11.4 :: float8)))
 -}
-dateTrunc ::
-  '[ null 'PGtext, null 'PGtimestamp ] ---> null 'PGtimestamp
-dateTrunc = unsafeFunctionN "date_trunc"
+dateTrunc
+  :: time `In` '[ 'PGtimestamp, 'PGtimestamptz ]
+  => TimeUnit -> null time --> null time
+dateTrunc tUnit args = unsafeFunctionN "date_trunc" (timeUnitExpr *: args)
+  where
+  timeUnitExpr :: forall grp lat with db params from null0.
+    Expression grp lat with db params from (null0 'PGtext)
+  timeUnitExpr = UnsafeExpression . singleQuotedUtf8 . renderSQL $ tUnit
+
+type family PGAtTimeZone ty where
+  PGAtTimeZone 'PGtimestamptz = 'PGtimestamp
+  PGAtTimeZone 'PGtimestamp = 'PGtimestamptz
+  PGAtTimeZone 'PGtimetz = 'PGtimetz
+  PGAtTimeZone pg = TypeError
+    ( 'Text "Squeal type error: AT TIME ZONE cannot be applie to "
+      ':<>: 'ShowType pg )
 
 {-|
-Truncate a timestamptz with the specified precision
-
->>> printSQL $ dateTrunctz ("day" *: makeTimestamptz (2010 :* 5 :* 6 :* 14 :* 45 *: 11.4))
-date_trunc((E'day' :: text), make_timestamptz((2010 :: int4), (5 :: int4), (6 :: int4), (14 :: int4), (45 :: int4), (11.4 :: float8)))
--}
-dateTrunctz ::
-  '[ null 'PGtext, null 'PGtimestamptz ] ---> null 'PGtimestamptz
-dateTrunctz = unsafeFunctionN "date_trunc"
-
-{-|
-Convert a timestamptz to a different timezone using an interval offset
-
->>>  printSQL $ (makeTimestamptz (2015 :* 6 :* 9 :* 3 :* 45 *: 11.4)) `atTimeZonetz` (interval_ (-1) Hours )
-(make_timestamptz((2015 :: int4), (6 :: int4), (9 :: int4), (3 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE (INTERVAL '-1.000 hours'))
--}
-atTimeZonetz ::
-  Operator (null 'PGtimestamptz) (null 'PGinterval) (null 'PGtimestamp)
-atTimeZonetz = unsafeBinaryOp "AT TIME ZONE"
-
-{-|
-Convert a timestamp to a different timezone using an interval offset
+Convert a timestamp, timestamp with time zone, or time of day with timezone to a different timezone using an interval offset
 
 >>> printSQL $ (makeTimestamp (2009 :* 7 :* 22 :* 19 :* 45 *: 11.4)) `atTimeZone` (interval_ 8 Hours)
 (make_timestamp((2009 :: int4), (7 :: int4), (22 :: int4), (19 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE (INTERVAL '8.000 hours'))
+
+>>> printSQL $ (makeTimestamptz (2015 :* 9 :* 15 :* 4 :* 45 *: 11.4)) `atTimeZone` (interval_ 1 Days)
+(make_timestamptz((2015 :: int4), (9 :: int4), (15 :: int4), (4 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE (INTERVAL '1.000 days'))
 -}
 atTimeZone ::
-  Operator (null 'PGtimestamp) (null 'PGinterval) (null 'PGtimestamptz)
+  Operator (null time) (null 'PGinterval) (null (PGAtTimeZone time))
```

Comment:
why is the right operand an `interval` rather than `text`?

Reply:
It can be specified as a specific timezone, with text, or as an interval. It would probably be better for the function to accept both intervals AND text representing the specific time zones.

[postgresql docs AT TIME ZONE](https://www.postgresql.org/docs/12/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT)

> In these expressions, the desired time zone zone can be specified either as a text string (e.g., 'America/Los_Angeles') or as an interval (e.g., INTERVAL '-08:00').

demonstration:
```
[db] # select make_timestamp((2009 :: int4), (7 :: int4), (22 :: int4), (19 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE  (INTERVAL '8.000 hours');
         timezone
--------------------------
 2009-07-22 11:45:11.4+00
(1 row)
[db] # select make_timestamp((2009 :: int4), (7 :: int4), (22 :: int4), (19 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE 'EST';
         timezone
--------------------------
 2009-07-23 00:45:11.4+00
```

Reply:
Maybe we just want to do Text, and not Intervals, postgresql won't allow intervals that are larger than 1 day?

Reply:
🤷

Reply:
`zone `In` '[ 'PGtext, 'PGinterval]`?